tool/tsh/db.go

```diff
@@ -834,59 +792,44 @@ func onDatabaseConnect(cf *CLIConf) error {
 	return nil
 }
 
-// getDatabaseInfo fetches information about the database from tsh profile is DB is active in profile. Otherwise,
-// the ListDatabases endpoint is called.
-func getDatabaseInfo(cf *CLIConf, tc *client.TeleportClient) (*tlsca.RouteToDatabase, types.Database, error) {
-	database, err := pickActiveDatabase(cf)
-	if err == nil {
-		switch database.Protocol {
-		case defaults.ProtocolCassandra:
-			// Cassandra CLI connection require database resource to determine
-			// if the target database is AWS hosted in order to skip the password prompt.
-		default:
-			return database, nil, nil
-		}
-	}
-	if err != nil && !trace.IsNotFound(err) {
-		return nil, nil, trace.Wrap(err)
-	}
-
-	dbService := cf.DatabaseService
-	username := cf.DatabaseUser
-	databaseName := cf.DatabaseName
-	if database != nil {
-		if dbService == "" {
-			dbService = database.ServiceName
-		}
-		if username == "" {
-			username = database.Username
-		}
-		if databaseName == "" {
-			databaseName = database.Database
-		}
-	}
-
-	db, err := getDatabase(cf, tc, dbService)
-	if err != nil {
-		return nil, nil, trace.Wrap(err)
+// getDatabaseInfo fetches information about the database from tsh profile if DB
+// is active in profile. Otherwise, the ListDatabases endpoint is called.
+func getDatabaseInfo(cf *CLIConf, tc *client.TeleportClient) (*databaseInfo, error) {
+	if route, err := pickActiveDatabase(cf); err == nil {
+		return newDatabaseInfo(cf, tc, *route)
+	} else if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
 	}
+	return newDatabaseInfo(cf, tc, tlsca.RouteToDatabase{
+		ServiceName: cf.DatabaseService,
+		Username:    cf.DatabaseUser,
+		Database:    cf.DatabaseName,
+	})
+}
 
-	return &tlsca.RouteToDatabase{
-		ServiceName: db.GetName(),
-		Protocol:    db.GetProtocol(),
-		Username:    username,
-		Database:    databaseName,
-	}, db, nil
+// databaseInfo wraps a RouteToDatabase and the corresponding database.
+// Its purpose is to prevent repeated fetches of the same database, by lazily
+// fetching and caching the database for use as needed.
+type databaseInfo struct {
+	tlsca.RouteToDatabase
+	// _database is the database corresponding to the db route and may be
+	// nil, so use GetDatabase instead of accessing it directly.
+	_database types.Database
```

Comment:
Is it common in Go for field names to start with an underscore? `rg "^\s+\b_[a-z]" -t go` returns only a couple of examples in our codebase. Wouldn't something like `databaseCache` also work here?

I'm not necessarily opposed to it, it just looks a bit out of place when the language already has a way to control the visibility of fields.

Reply:
I wanted to emphasize that the private field shouldn't be accessed directly. Since this type is declared in the same package as tsh, it can be accessed by anything in tsh even though it's not exported 😅